src/pathfinder/yapf/yapf_ship_regions.cpp

```diff
@@ -0,0 +1,314 @@
+		{
+			Node &node = Yapf().CreateNewNode();
+			node.Set(&old_node, water_region_patch);
+			Yapf().AddNewNode(node, TrackFollower{});
+		};
+		VisitWaterRegionPatchNeighbors(old_node.m_key.m_water_region_patch, visitFunc);
+	}
+
+	inline char TransportTypeChar() const { return '^'; }
+
+	static std::vector<WaterRegionPatchDesc> FindWaterRegionPath(const Ship *v, TileIndex start_tile, int max_returned_path_length)
+	{
+		const WaterRegionPatchDesc start_water_region_patch = GetWaterRegionPatchInfo(start_tile);
+
+		/* We reserve 4 nodes (patches) per water region. The vast majority of water regions have 1 or 2 regions so this should be a pretty
+		 * safe limit. We cap the limit at 65536 which is at a region size of 16x16 is equivalent to one node per region for a 4096x4096 map. */
+		Tpf pf(std::min(static_cast<int>(Map::Size() * NODES_PER_REGION) / WATER_REGION_NUMBER_OF_TILES, MAX_NUMBER_OF_NODES));
+		pf.SetDestination(start_water_region_patch);
+
+		if (v->current_order.IsType(OT_GOTO_STATION)) {
+			DestinationID station_id = v->current_order.GetDestination();
+			const BaseStation *station = BaseStation::Get(station_id);
+			TileArea tile_area;
+			station->GetTileArea(&tile_area, STATION_DOCK);
+			for (const auto &tile : tile_area) {
+				if (IsDockingTile(tile) && IsShipDestinationTile(tile, station_id)) {
+					pf.AddOrigin(GetWaterRegionPatchInfo(tile));
+				}
+			}
+		} else {
+			TileIndex tile = v->dest_tile;
+			pf.AddOrigin(GetWaterRegionPatchInfo(tile));
+		}
+
+		/* If origin and destination are the same we simply return that water patch. */
+		std::vector<WaterRegionPatchDesc> path = { start_water_region_patch };
+		path.reserve(max_returned_path_length);
+		if (pf.HasOrigin(start_water_region_patch)) return path;
+
+		/* Find best path. */
+		if (!pf.FindPath(v)) return {}; // Path not found.
+
+		Node *node = pf.GetBestNode();
+		for (int i = 0; i < max_returned_path_length - 1; ++i) {
+			if (node != nullptr) {
+				node = node->m_parent;
+				if (node != nullptr) path.push_back(node->m_key.m_water_region_patch);
+			}
+		}
+
+		assert(!path.empty());
+		return path;
+	}
+};
+
+/** Cost Provider of YAPF for water regions. */
+template <class Types>
+class CYapfCostRegionT
+{
+public:
+	typedef typename Types::Tpf Tpf;              ///< The pathfinder class (derived from THIS class).
+	typedef typename Types::TrackFollower TrackFollower;
+	typedef typename Types::NodeList::Titem Node; ///< This will be our node type.
+	typedef typename Node::Key Key;               ///< Key to hash tables.
+
+protected:
+	/** To access inherited path finder. */
+	Tpf &Yapf() { return *static_cast<Tpf*>(this); }
+
+public:
+	/**
+	 * Called by YAPF to calculate the cost from the origin to the given node.
+	 * Calculates only the cost of given node, adds it to the parent node cost
+	 * and stores the result into Node::m_cost member.
+	 */
+	inline bool PfCalcCost(Node &n, const TrackFollower */*tf*/)
```

Comment:
What happened here? GitHub shows it as red .. :D

Reply:
Done!